src/gc-heap-snapshot.cpp

```diff
@@ -106,15 +106,22 @@ struct StringTable {
 };
 
 struct HeapSnapshot {
-    vector<Node> nodes;
-    // edges are stored on each from_node
-
     StringTable names;
     StringTable node_types;
     StringTable edge_types;
     DenseMap<void *, size_t> node_ptr_to_index_map;
```

Comment:
I think you can also avoid creating this table in the serializer process by making it an identity map instead (when used for edges) and streaming out the value of the original pointer also on each node. Then in the later process, you just need to build this map from pointer->number to satisfy converting it into the Javascript format. This dict is also used to serialize the representation/name of the object exactly once, but that can instead be satisfied by knowing the GC will visit each object exactly once as a "from" node when marking all of the out-refs.

Reply:
Yeah, great points!
You're right, i think it should be a pattern of N edges from node 1 then M edges from node 2, etc. 👍 So we should be able to account for that in the second half of your comment.

I hadn't noticed this map yet in my rush job to stream the data here. I think you're right that we should fix this too 👍 👍 

Thanks!

Reply:
We considered this, but didn't address it in the new PR https://github.com/JuliaLang/julia/pull/52854 since it requires reprocessing of the nodes/edges and some complex logic to merge duplicate nodes and outgoing edges and it could only be an issue for a huge snapshot. We like to leave it for future work.